KLEE floating-point support.

qrort opened the pull request:
The implementation is based on KLEE-float.

Supported llvm math intrinsic functions related to floating-point.

Added ENABLE_FP and FP_RUNTIME options to CMake.

Provided an ability to switch between Z3Builder implementations.

Building KLEE without -DCMAKE_ENABLE_FLOATING_POINT does not change the core.

Thank you for contributing to KLEE.  We are looking forward to reviewing your PR.  However, given the small number of active reviewers and our limited time, it might take a while to do so.  We aim to get back to each PR within one month, and often do so within one week. 

To help expedite the review please ensure the following, by adding an "x" for each completed item:

- [ ] The PR addresses a single issue.  In other words, if some parts of a PR could form another independent PR, you should break this PR into multiple smaller PRs.
- [x] There are no unnecessary commits. For instance, commits that fix issues with a previous commit in this PR are unnecessary and should be removed (you can find [documentation on squashing commits here](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request#squash-your-changes)).
- [ ] Larger PRs are divided into a logical sequence of commits.
- [x] Each commit has a meaningful message documenting what it does.
- [ ] The code is commented.  In particular, newly added classes and functions should be documented.
- [x] The patch is formatted via  [clang-format](https://clang.llvm.org/docs/ClangFormat.html) (see also [git-clang-format](https://raw.githubusercontent.com/llvm/llvm-project/master/clang/tools/clang-format/git-clang-format) for Git integration).  Please only format the patch itself and code surrounding the patch, not entire files.  Divergences from clang-formatting are only rarely accepted, and only if they clearly improve code readability.
- [x] Add test cases exercising the code you added or modified.  We expect [system and/or unit test cases](https://klee.github.io/docs/developers-guide/#regression-testing-framework) for all non-trivial changes.  After you submit your PR, you will be able to see a [Codecov report](https://docs.codecov.io/docs/pull-request-comments) telling you which parts of your patch are not covered by the regression test suite.  You will also be able to see if the Github Actions CI and Cirrus CI tests have passed.  If they don't, you should examine the failures and address them before the PR can be reviewed. 
- [x] Spellcheck all messages added to the codebase, all comments, as well as commit messages.

251 committed:
differentiate between partial and completed paths in summary and fix paths stats when not dumping states

251 committed:
posix runtime: add malloc checks

jbuening committed:
add ifdefs for C++ exception handling

restoring old behavior without EH support

jbuening committed:
renaming 'libcxx' -> 'libc++'

jbuening committed:
add klee_messages for C++ exception handling support

otherwise, it is hard to tell whether EH support is available or not

ccadar committed:
Refactored MaxStatis*Pct conditions into a separate function.

ccadar committed:
Added -max-static-pct-check-delay to replace the hardcoded delay after which the MaxStatic*Pct checks are performed.

ccadar committed:
Added a warning when forking is skipped due to MaxStatic*Pct limits being reached

ccadar committed:
Test for -max-static-fork-pct

jbuening committed:
Executor: remove obsolete special case for __cxa_{re,}throw

__cxa_throw and __cxa_rethrow were not handled by special function handlers in
the final version of #966 (which introduced support for C++ exception handling)

ccadar committed:
Improved PR template by making it more succinct.

jiseongg committed:
Add cmake custom target `uninstall`

`make uninstall` is enabled.
Without this, uninstalling was done manually with `rm` command.

Commit:
[Z3] Handle the case when interruption caught by Z3

Commit:
KLEE floating-point support.

The implementation is based on KLEE-float.

Supported llvm math intrinsic functions related to floating-point.

Added ENABLE_FP and FP_RUNTIME options to CMake.

Provided an ability to switch between Z3Builder implementations.

Building KLEE without -DCMAKE_ENABLE_FLOATING_POINT does not change the core.

qrort commented:
The patch was created as part of Software Analysis project in Huawei Russian Research Institute.

operasfantom commented on include/klee/Support/ModuleUtil.h:
```diff
@@ -39,6 +39,59 @@ std::unique_ptr<llvm::Module>
+        {"llvm."#from".f32", #to"f"},     \
+        {"llvm."#from".f64", #to},        \
+        {"llvm."#from".f80", #to"l"}      \
+
```
What about the intrinsics ending with ".f128"? For example, call of `__builtin_copysignf128` in C/C++ code generates call of `@llvm.copysign.f128` in IR code

jirislaby committed:
runtime/POSIX: fix failures with glibc-2.33

commit 8ed005daf0ab of glibc-2.33 (Remove stat wrapper functions, move
them to exported symbols) removed renames of `__fxstat`, `__xstat`, and
`__lxstat` to `__fxstat64`, `__xstat64`, and `__lxstat64`, respectively.

But we relied on the renames to build `fd_64.c` properly. With glibc
2.33, we now see link failures of the POSIX runtime:
error: Linking globals named '__xstat': symbol multiply defined!

Rename the functions using `__REDIRECT_NTH` in the code as
`__USE_FILE_OFFSET64` case (which we define at the top of the file by
`#define _FILE_OFFSET_BITS 64`) did exactly that.

Fixes #1384.

operasfantom commented on lib/Module/IntrinsicCleaner.cpp:
```diff
@@ -70,7 +70,39 @@ bool IntrinsicCleanerPass::runOnBasicBlock(BasicBlock &b, Module &M) {
-        break;
+      case Intrinsic::maxnum:
+      case Intrinsic::minnum:
+      case Intrinsic::fma:
```
Have you modelled "fma"? I couldn't find that

251 committed:
tests: add support for LLVM 11.1

251 committed:
test: count paths with -dump-states-on-halt=false

lzaoral committed:
cmake: Fix warning about implicit type conversion

Fixes:
CMake Warning (dev) at CMakeLists.txt:478 (set):
  implicitly converting 'String' to 'STRING' type.
This warning is for project developers.  Use -Wno-dev to suppress it.

ccadar committed:
Refactored maxStaticPctChecks into a sequence of conditions.

ccadar committed:
Replaced the time-based delay after which the max-static-*-pct checks are performed with one expressed in terms of number of forks.

251 committed:
tests: adjust to new summary output

lzaoral committed:
cmake: Fix CMP0026 policy deprecation warning

CMake 3.19+ started to issue warnings if this policy is set to OLD:

CMake Deprecation Warning at test/CMakeLists.txt:143 (cmake_policy):
  The OLD behavior for policy CMP0026 will be removed from a future version
  of CMake.

  The cmake-policies(7) manual explains that the OLD behaviors of all
  policies are deprecated and that a policy should be set to OLD only under
  specific short-term circumstances.  Projects should be ported to the NEW
  behavior and not rely on setting a policy to OLD.

lzaoral committed:
tests: Do not add klee tools to PATH in lit

It may happen that some older instance of klee is already present
in PATH.  All tests that call plain klee instead of %klee may use
it and then unexpectedly fail.

This commit will make all tests that rely on klee tools being
explicitly in PATH fail in our CI.  From now on, only LLVM tools,
FileCheck and not will be in lit's PATH.

lzaoral committed:
tests: Invoke tools through their corresponding macros

lzaoral committed:
klee-replay: Fix -Wformat-truncation warning

Increase the size of the buffer to PATH_MAX in create_link as that is the
maximal possible length of fname and check whether output truncation occurred.

Fixes:
tools/klee-replay/file-creator.c: In function 'create_file':
tools/klee-replay/file-creator.c:55:31: warning: '%s' directive output may be truncated writing up to 4095 bytes into a region of size 64 [-Wformat-truncation=]
   55 |   snprintf(buf, sizeof(buf), "%s.lnk", fname);
      |                               ^~
......
  344 |   target = tmpname;
      |            ~~~~~~~
In file included from /usr/include/stdio.h:866,
                 from tools/klee-replay/file-creator.c:16:
/usr/include/bits/stdio2.h:70:10: note: '__snprintf_chk' output between 5 and 4100 bytes into a destination of size 64
   70 |   return __builtin___snprintf_chk (__s, __n, __USE_FORTIFY_LEVEL - 1,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   71 |        __bos (__s), __fmt, __va_arg_pack ());
      |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

lzaoral committed:
test/Concrete: Use Python 3 in ConcreteTest.py explicitly

Some Linux distributions, e.g. Fedora, do not install the unversioned Python
binary by default and Python 2 has been dead for more than a year.

jordr committed:
Fix erroneous klee-stats legend for --print-all

Only absolute times were displayed, and were marked as %.
Fixes CexCacheTime, ForkTime and ResolveTime columns.

251 committed:
cirrus: update FreeBSD python dependencies

jbuening committed:
extend function pointer test

Part of the test was already disabled in the initial checkin.
However, we do support function pointers if they are restricted to
one or more possible values.

jbuening committed:
allocate memory objects for functions

Before, we reused the llvm::Function* value in the target program,
even though it stems from KLEE's own address space. This leads to
non-deterministic function pointers, even with --allocate-determ.
This issue was identified in the MoKLEE paper. Now, we allocate a
memory object per function, for its (potentially) deterministic
address. Mapping this address back to llvm::Functions is done by
the legalFunctions map.

Also, pointer width now depends on the target, not the host.

Commit:
KLEE floating-point support.

The implementation is based on KLEE-float.

Supported llvm math intrinsic functions related to floating-point.

Added ENABLE_FP and FP_RUNTIME options to CMake.

Provided an ability to switch between Z3Builder implementations.

Building KLEE without -DCMAKE_ENABLE_FLOATING_POINT does not change the core.

ccadar commented:
@qrort Thanks for this contribution!  It would be of course great to have this in the mainline, but we need to integrate it more incrementally.  Would you be interested in working on such a step-by-step integration?

Commit:
fixed tests after 2.2 rebase

Commit:
fpclassify rewrited without fork

qrort commented:
@ccadar Yes, I would really like to eventually integrate the patch into KLEE! I have noticed several CMake configuration bugs in the MR which I have not addressed and applied to the patch yet. I have also done some experiments on Coreutils and some open-source GitHub libraries which show the patch applicability. We can communicate here or via emails; I have already sent you a couple throughout the year.

ccadar commented:
@qrort Thanks for the update.  Both email and GitHub communication is fine with me.  I saw this useful PR opened for a while now, and I wanted to see what is the status with it.  BTW, in case you haven't attended: at the recent KLEE workshop, FP support in KLEE (and the KLEE-Float extension) came up several times.

operasfantom commented:
That would be unfortunate to miss this great extension which is supposed to greatly increase the coverage and error detection. Do you have any plans with respect to extension's support?

danielschemmel commented:
@qrort you may also wish to answer the open review questions.

MartinNowack commented:
@operasfantom Thanks a lot for this PR. In it's current shape, this PR requires substantial effort to rebase it to the most recent version of KLEE with the time passed. I'm going to close this PR. If you are still interested to upstream this feature, feel free to re-open this PR or open a new one.

ccadar commented:
I have added this to the list of possible extensions: https://github.com/klee/klee/projects/4

floatshadow commented:
@ccadar Hi! May I ask about the progress of floating point support of KLEE? I found it removed from the list of possible extensions.

ccadar commented:
@floatshadow there are no updates here, and no plans at the moment to add FP support to mainline KLEE. 
The project was unintentionally moved from the list of extensions when GitHub retired support for "classic" projects.
